### PR TITLE
Implement ConsumerStateTable/ProducerStateTable which aggregate the table actions

### PR DESCRIPTION
**Pull request by qiluo-msft**
And a lot of refactors.

### CODE REVIEW

File: common/consumerstatetable.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            multi();
+            enqueue(std::string("SCARD ") + getKeySetName(), REDIS_REPLY_INTEGER);
+            subscribe(m_db, getChannelName());
+            bool succ = exec();
+            if (succ) break;
+        }
+
+        long long int len = queueResultsFront()->integer;
+        setQueueLength(len);
+    }
```
Comment (kcudnik) [inline common/consumerstatetable.h]: why put code inside header? move to cpp ?

File: common/rediscommand.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    size_t length() const
+    {
+        return strlen(temp);
+    }
+    
+private:
+    char *temp;
+};
+
+}
```
Comment (kcudnik) [inline common/rediscommand.h]: we could move redisclient.cpp from sairedis to this

File: common/redisreply.cpp [hunk 3]
```diff
@@ -77,6 +72,33 @@ void RedisReply::checkStatus(char *status)
...
 }
 
+void RedisReply::checkReply()
+{
+    if (!m_reply)
+    {
+        throw system_error(make_error_code(errc::not_enough_memory),
+                           "Memory exception, reply is null");
+    }
+
```
Comment (stcheng) [inline common/redisreply.cpp]: empty spaces

File: common/table.cpp [hunk 4]
```diff
@@ -4,49 +4,16 @@
 #include "common/table.h"
 #include "common/logger.h"
 #include "common/redisreply.h"
+#include "common/rediscommand.h"
 
 using namespace std;
```
Comment (stcheng) [inline common/table.cpp]: as long as there's std namepsace, std:: could be removed in the file.

File: common/table.h [hunk 5]
```diff
@@ -19,77 +20,86 @@ typedef std::tuple<std::string, std::string, std::vector<FieldValueTuple> > KeyO
...
-    Table(DBConnector *db, std::string tableName);
+    TableBase(std::string tableName) : m_tableName(tableName) { }
 
-    /* Read a value from the DB directly */
-    /* Returns false if the key doesn't exists */
-    bool get(std::string key, std::vector<FieldValueTuple> &values);
+    std::string getTableName() const { return m_tableName; }
+    
+    /* Return the actual key name as a comibation of tableName:key */
+    std::string getKeyName(std::string key) { assert(!key.empty()); return m_tableName + ':' + key; }
```
Comment (stcheng) [inline common/table.h]: i think there's a case that the key is empty and we only have the table name. in this case the table only have one entry and we treat this table itself as a key field value entry.

File: common/table.h [hunk 6]
```diff
@@ -19,77 +20,86 @@ typedef std::tuple<std::string, std::string, std::vector<FieldValueTuple> > KeyO
...
-    void delField(std::string key, std::string field);
+                     std::string op = "") = 0;
+    /* Delete an entry in the table */
+    virtual void del(std::string key, std::string op = "") = 0;
+    
+};
 
-    virtual ~Table();
-    std::string getTableName() const;
+class TableEntryPopable {
```
Comment (stcheng) [inline common/table.h]: Poppable

File: common/table.h [hunk 7]
```diff
@@ -19,77 +20,86 @@ typedef std::tuple<std::string, std::string, std::vector<FieldValueTuple> > KeyO
...
+public:
+    virtual ~TableEntryPopable() { }
+    
+    /* Pop an written action (set or del) on the table */
```
Comment (stcheng) [inline common/table.h]: a written action? or an action?

File: common/table.h [hunk 8]
```diff
@@ -19,77 +20,86 @@ typedef std::tuple<std::string, std::string, std::vector<FieldValueTuple> > KeyO
...
-                                  const std::string& field);
+class Table : public RedisTransactioner, public TableBase, public TableEntryEnumerable {
+public:
+    Table(DBConnector *db, std::string tableName);
+    virtual ~Table() { }
 
-    /* Format HDEL key field command */
-    static std::string formatHDEL(const std::string& key,
-                                  const std::string& field);
+    /* Set an entry in the DB directly (op not in used) */
```
Comment (stcheng) [inline common/table.h]: not in use

File: common/table.h [hunk 9]
```diff
@@ -19,77 +20,86 @@ typedef std::tuple<std::string, std::string, std::vector<FieldValueTuple> > KeyO
...
+    virtual void set(std::string key, std::vector<FieldValueTuple> &values,
+                     std::string op = "");
+    /* Delete an entry in the DB directly (op not in used) */
```
Comment (stcheng) [inline common/table.h]: not in use

File: tests/redis_state_ut.cpp [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    int ret, i = 0;
+    KeyOpFieldsValuesTuple kco;
+
+    cs.addSelectable(&c);
+    while ((ret = cs.select(&selectcs, &tmpfd)) == Select::OBJECT)
+    {
+        c.pop(kco);
+        if (kfvOp(kco) == "SET")
+        {
+            numberOfKeysSet++;
```
Comment (stcheng) [inline tests/redis_state_ut.cpp]: is this value used anywhere?

File: tests/redis_state_ut.cpp [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        Selectable *is;
+        int fd;
+
+        ret = cs.select(&is, &fd);
+        EXPECT_EQ(ret, Select::OBJECT);
+
+        ((ConsumerStateTable *)is)->pop(kco);
+        if (kfvOp(kco) == "SET")
+        {
+            numberOfKeysSet++;
```
Comment (stcheng) [inline tests/redis_state_ut.cpp]: is this value used anywhere? compare if it equals the numberOfKeysDeleted?

File: common/table.h (hunk 5 above)
Comment (qiluo-msft) [inline common/table.h]: Could you find this case? I have grep but not found any.

---
In reply to: [85476261](https://github.com/Azure/sonic-swss-common/pull/55#discussion_r85476261) [](ancestors = 85476261)

Comment (stcheng) [inline common/table.h]: `                            /* Add port to port list */
                            m_portList[alias] = p;
                            /* Add port name map to counter table */
                            std::stringstream ss;
                            ss << hex << p.m_port_id;
                            FieldValueTuple tuple(p.m_alias, ss.str());
                            vector<FieldValueTuple> vector;
                            vector.push_back(tuple);
                            m_counterTable->set("", vector);
`

File: common/consumerstatetable.h (hunk 1 above)
Comment (qiluo-msft) [inline common/consumerstatetable.h]: For easy code management, keep it here. I would like to move if you still prefer.

---
In reply to: [85452806](https://github.com/Azure/sonic-swss-common/pull/55#discussion_r85452806) [](ancestors = 85452806)

File: common/consumerstatetable.cpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    if (r.getContext()->type == REDIS_REPLY_NIL)
+    {
+        kfvKey(kco).clear();
+        kfvOp(kco).clear();
+        return;
+    }
+
+    assert(ctx->type == REDIS_REPLY_ARRAY);
+    assert(ctx->elements == 2);
+    assert(ctx->element[0]->type == REDIS_REPLY_STRING);
```
Comment (lguohan) [inline common/consumerstatetable.cpp]: do we need these asserts?

File: common/producerstatetable.cpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    : TableName_KeySet(tableName)
+    , m_db(db)
+{
+}
+
+void ProducerStateTable::set(std::string key, std::vector<FieldValueTuple> &values,
+                 std::string op /*= SET_COMMAND*/)
+{
+    static std::string luaScript =
+        "redis.call('PUBLISH', KEYS[1], ARGV[1])\n"
```
Comment (lguohan) [inline common/producerstatetable.cpp]: it does not make logic difference, but it's better to put publish at the end.

one future optimization we can do is not to do publish every time, we can take look at the set, if the set is not empty, we do not need to do publish.

File: common/producerstatetable.cpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+    std::string command = osk.str() + osv.str();
+
+    RedisReply r(m_db, command, REDIS_REPLY_NIL);
+}
+
+void ProducerStateTable::del(std::string key, std::string op /*= DEL_COMMAND*/)
+{
+    static std::string luaScript =
+        "redis.call('PUBLISH', KEYS[1], ARGV[1])\n"
```
Comment (lguohan) [inline common/producerstatetable.cpp]: move publish to the end, align with other class for the style.

File: tests/redis_state_ut.cpp [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        int maxNumOfFields = getMaxFields(i);
+        for (int j = 0; j < maxNumOfFields; j++)
+        {
+            FieldValueTuple t(field(j), value(j));
+            fields.push_back(t);
+        }
+        if ((i % 100) == 0)
+            cout << "+" << flush;
+
+        p.set(key(i), fields);
```
Comment (lguohan) [inline tests/redis_state_ut.cpp]: >p.set(key(i), fields); [](start = 8, length = 22)

all of these keys are different. 

I cannot find test that specifically test the duplicate keys for the state producer class. is there any? can you point out?

File: common/consumerstatetable.cpp (hunk 12 above)
Comment (qiluo-msft) [inline common/consumerstatetable.cpp]: Yes, they are valid assertions.

---
In reply to: [86660211](https://github.com/Azure/sonic-swss-common/pull/55#discussion_r86660211) [](ancestors = 86660211)

File: tests/redis_state_ut.cpp (hunk 15 above)
Comment (qiluo-msft) [inline tests/redis_state_ut.cpp]: Will add

---
In reply to: [86660316](https://github.com/Azure/sonic-swss-common/pull/55#discussion_r86660316) [](ancestors = 86660316)

**Comment by lguohan** [comment]
build it

**Comment by lguohan** [comment]
test

File: common/consumerstatetable.cpp [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+#include <vector>
+#include <limits>
+#include <hiredis/hiredis.h>
+#include "dbconnector.h"
+#include "table.h"
+#include "selectable.h"
+#include "redisselect.h"
+#include "redisapi.h"
+#include "consumerstatetable.h"
+
```
Comment (stcheng) [inline common/consumerstatetable.cpp]: using namespace std;
the other files use it.

File: common/producerstatetable.cpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
+#include <stdlib.h>
+#include <tuple>
+#include <sstream>
+#include "redisreply.h"
+#include "table.h"
+#include "redisapi.h"
+#include "producerstatetable.h"
+
```
Comment (stcheng) [inline common/producerstatetable.cpp]: using namepsace std;

File: common/rediscommand.h [hunk 18]
```diff
@@ -0,0 +1,92 @@
+#pragma once
```
Comment (stcheng) [inline common/rediscommand.h]: some places use pragma once, some places use ifdef/define, better to unify them.

File: common/table.cpp [hunk 19]
```diff
@@ -339,28 +118,16 @@ void Table::dump(TableDump& tableDump)
...
+    static std::string sha = loadRedisScript(m_db, luaScript);
+    
     SWSS_LOG_TIMER("getting");
-
-    char *temp;
-
-    int len = redisFormatCommand(
-            &temp,
-            "EVALSHA %s 1 %s ''",
+
```
Comment (stcheng) [inline common/table.cpp]: empty spaces

File: common/table.cpp [hunk 20]
```diff
@@ -339,28 +118,16 @@ void Table::dump(TableDump& tableDump)
 
         "return cjson.encode(res)\n";
 
-    static std::string sha = scriptLoad(luaScript);
-
+    static std::string sha = loadRedisScript(m_db, luaScript);
+
```
Comment (stcheng) [inline common/table.cpp]: empty spaces

File: common/consumerstatetable.cpp (hunk 12 above)
Comment (qiluo-msft) [inline common/consumerstatetable.cpp]: Assertions are always redundant. They are here to explain that below statements are safe.

---
In reply to: [86672982](https://github.com/Azure/sonic-swss-common/pull/55#discussion_r86672982) [](ancestors = 86672982)

**Pull request merged**